### ISSUE TITLE
Solvers do not preserve the real type of their parameters

### ISSUE THREAD

**Issue by tcojean** [ginkgo-project/ginkgo#423]
This is a problem found by @nbeams. The question is how can we appropriately solve it?

The implementation of the current solvers, for example in `CG` initializes the temporary variables in the following way:
```c++
    using Vector = matrix::Dense<ValueType>;
    // ...
    auto dense_b = as<const Vector>(b);
    auto dense_x = as<Vector>(x);
    // ...
    auto p = Vector::create_with_config_of(dense_b);
    auto q = Vector::create_with_config_of(dense_b);
```

The problem concerns if we allow (which we do) to derive from Dense, and someone passes in as `b` a derived class of `Dense<...>`, then all the temporary variables will have the plain type `Dense<...>` instead of the type of `b`, which could (would) threaten the correct execution of the algorithm.

**Comment by thoasm** [comment]
Thanks for finding this @nbeams. I actually did not think of the other vectors we instantiate at the start of each solver.
Since there is no nice way of getting the actual type from the vector `x`, I currently only see one of these options:
1. use `clone()` instead of `create_with_config_of` (will cost a bit of performance since we do an additional copy and likely requires to override the `clone_imp`)
2. make a virtual function for `create_with_config_of`, so it calls the proper derived function
3. Add a `clone_allocation` (or a better name) to the `LinOp` interface, which is supposed to allocate the appropriate memory without copying.

I would go for option (2.) because we would prevent the unnecessary copying from option (1.). Option (3.) might also be quite complicated, because I would not know how to do the `clone_allocation` for solvers (for example do we just allocate the associated matrix, or do we properly fill it?).

I appreciate any input, for example if you know additional suggestions or feel like we should go for another option.

**Comment by tcojean** [comment]
Thanks for writing this out @thoasm. I think option 2. is the most sensible, with a `create_with_config_of` and an `_impl` version of it, probably.

**Comment by nbeams** [comment]
I agree option (2) seems like the best choice, but we'd still need to do a little more rearranging beyond letting classes override `create_with_config_of`, right?  In the cg solver, at least, it calls `as<Dense>(b)` first, then passes the result of that cast to `create_with_config_of`, meaning any classes derived from Dense are still treated like base Dense when making the local vectors, no matter what the derived class overrides.

Would you still want to use the `as<Dense>` to confirm that the arguments are either Dense or something derived from Dense, but not use the resulting pointer (assuming a successful cast) for anything?  Or would that not be necessary because only Dense and its derived classes would have the `create_with_config_of` enabled, so any other argument type would cause an error?

**Comment by thoasm** [comment]
`create_with_config_of` will still return a `std::unique_ptr<Dense>`, we can't change that unfortunately (because we don't know the dynamic type during compile time). That means you can only access methods from `Dense`, and not from your derived class without additional casts.
However, that is perfectly fine for all existing code because they only use methods which are available in `Dense`. If you need special handling (which you probably do) inside the `apply_impl()`, you need to `dynamic_cast` the parameters to the type you expect (I would go with using the `as<Derived>()`, so it throws if the type mismatches). But you don't need to do this for the `this` pointer in your `apply_impl()` since you already know it is the proper type.

I am fine with implementing the virtual function for the `create_with_config_of` (we need this either way at some point and it is really not a lot of code), unless you want to do that. Either way, I think a separate PR would be nice here since it is a feature / bug fix in its own.
In theory, it is just adding a protected virtual function (with a useful name like maybe `create_without_copy` or similar), which then will be called from the static function `create_with_config_of` (instead of calling the creating on its own).

**Comment by nbeams** [comment]
No, I mean even if we have a version of `create_with_config_of` that I can override, the cg code as currently written won't call my version (even if the original objects are of my derived type) because `create_with_config_of` is only called on the result of the as<Dense> cast. 

If I have to call `create_with_config_of` inside my matvec apply_impl to create wrapped MFEM vectors, I don't think it's any better than what I was doing before (having to create/destroy MFEM Vectors inside each matvec).

**Comment by thoasm** [comment]
Don't you also have a derived matrix for MFEM? I thought you had, in which case you can `dynamic_cast` to your derived Dense to see if it is there. If it is, do all the MFEM operations, otherwise, do the conversions that you are doing currently.

**Comment by thoasm** [comment]
It would call your version of `create_with_config_of` if you call a virtual function inside of it.
I should be pretty fast in writing the code, so I can show you in the PR what I mean. It should definitively create an object of your derived `Dense` format, just the static type will always be `Dense` (which we can't change, at least in the solvers). If you call apply (or any virtual function you overwrote) on the newly generated object, it will definitively call the derived function and not the original `Dense` one.

**Comment by nbeams** [comment]
I'm not sure I understand your comment about how I can use the matrix.  The matrix wrapper class is just derived from LinOp, not any matrix type.  I was wanting to make a general MFEM wrapper that would work when using one of MFEM's assembly modes that does not build and store a matrix, so all we have available to us is the action of applying the operator, which I define in apply_impl of my LinOp class.

**Comment by thoasm** [comment]
Since your MFEM wrapper is a `LinOp`, you can check inside there if you can `dynamic_cast` your `x` and `b` into your derived class, or not. We do something similar in the [CSR implementation](https://github.com/ginkgo-project/ginkgo/blob/develop/core/matrix/csr.cpp#L89) to check if the `x` is another CSR matrix (in which case we do an SpGEMM), or not (just a regular SpMV).

In your case, I would check if it is your own format (in which case you can proceed without any conversions), or not (where you can decide to either throw or do the conversions).

I hope that the PR helps to show what I mean and clarifies things I did not explain properly.

**Comment by thoasm** [comment]
@nbeams If you have questions about the fix in PR #424 or how to use it, feel free to re-open this issue or write your question in slack.
